Update nullable reference types for C# 10

BillWagner opened the pull request:
There are a number of changes here:

**First two commits:** I reworked the conceptual article to focus on the compiler's static analysis instead of the addition of nullable reference types as a type. That addresses many of the misconceptions that led to the various issues fixed in this PR.

Fixes #21716

- Beginning in .NET 6, new projects have `<nullable>enable</nullable>` on by default.
- Rework the conceptual article to focus more on the nullable analysis than on the new type annotations.

**Third commit:**

Fixes #24257

- Provide a more detailed description and recommendations on using the different settings for annotations and warnings in a nullable context.

**Fourth commit**

Fixes #25027

- Update the descriptions of the behavior for the annotation and warning contexts.


[Internal preview - nullable overview](https://review.docs.microsoft.com/en-us/dotnet/csharp/nullable-references?branch=pr-en-us-25883)

BillWagner committed:
Change focus from types to static analysis

Apply suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>

final edit

IEvangelist reviewed (commented):
I love the way this is shaping, here is some earlier feedback.

BillWagner committed:
Fixes for #24257

Provide a more detailed description and recommendations on using the different settings for annotations and warnings in a nullable context.

BillWagner committed:
Fixes #25027

Fix descriptions of the warnings and annotations settings for nullable contexts.

IEvangelist commented on docs/csharp/nullable-references.md:
```diff
@@ -2,75 +2,171 @@
+- Choose *annotations* to express your design intent before enabling warnings.
+- Choose *enabled* for new projects and active projects where you want to protect against null reference exceptions.
 
 The nullable annotation context and nullable warning context can be set for a project using the `Nullable` element in your *.csproj* file. This element configures how the compiler interprets the nullability of types and what warnings are generated. Valid settings are:
```
```suggestion
The nullable annotation context and nullable warning context can be set for a project using the `<Nullable>` element in your *.csproj* file. This element configures how the compiler interprets the nullability of types and what warnings are generated. Valid settings are:
```

IEvangelist commented on docs/csharp/nullable-references.md:
```diff
@@ -92,46 +204,13 @@ You can also use directives to set these same contexts anywhere in your project:
 ## Known pitfalls
 
-Arrays and structs that contain reference types are known pitfalls in nullable reference types feature.
+Arrays and structs that contain reference types are known pitfalls in nullable references and the static analysis that determines null safety.
```
How are the known pitfalls, what about them specifically? Like if they contain reference types, and one of the elements is `null`?

BillWagner replied:
Added an introductory statement here.

IEvangelist reviewed (approved):
LGTM, had some suggestions and questions. Otherwise feel free to :shipit:

BillWagner commented on docs/csharp/nullable-references.md:
````diff
@@ -2,75 +2,171 @@
+Consider the following example:
+
+```csharp
+string message = null;
+Console.WriteLine($"The length of the message is {message.Length}); // warning: dereference null.
+var originalMessage = message;
+message = "Hello, World!";
+Console.WriteLine($"The length of the message is {message.Length}); // No warning. Analysis determined "message" is not null.
+Console.WriteLine(originalMessage.Length); // warning!
````
I'd leave it as `string`. I haven't introduced the `string?` type yet.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>

BillWagner commented on docs/csharp/nullable-references.md:
```diff
@@ -2,75 +2,171 @@
+- Attributes that annotate APIs so that the flow analysis determines *null-state*.
+- Variable annotations that developers use to explicitly declare the intended *null-state* for a variable.
+
+The rest of this article describes how those three feature areas work to produce warnings when your code may be dereferencing a `null` value.
```
I added a definition for "dereferencing". I think it's a diversion to mention other ways to throw a `NullReferenceException`. (But you're right).

BillWagner commented on docs/csharp/nullable-references.md:
````diff
@@ -2,75 +2,171 @@
 ```
 
-## Nullability of types
+Nullable reference types and nullable value types provide a similar semantic concept: A variable can represent a value or object, or that variable may refer to nothing. When a variable refers to nothing, it's `null`. However, nullable reference types and nullable value types are implemented differently: nullable value types are implemented using <xref:System.Nullable%601?displayProperty=nameWithT ...
````
I was trying to dance around an online argument where someone asserted that "nullable value types can't be null because they are value types.". I stopped dancing around it. The spec has that a nullable value type whose `HasValue` property is `false` is equal to `null`.

BillWagner committed:
respond to feedback.

gewarren commented:
@BillWagner I think this introduced some bad lint.
